torch/csrc/jit/pickler.cpp

```diff
@@ -145,6 +148,24 @@ void Pickler::pushDevice(const IValue& ivalue) {
   }
 }
 
+void Pickler::pushRRef(const IValue& ivalue) {
+  auto rrefInterface = ivalue.toRRef();
+  auto rref = c10::static_intrusive_pointer_cast<distributed::rpc::RRef>(rrefInterface);
+  pushGlobal("torch", "rref");
+  auto& ctx = distributed::rpc::RRefContext::getInstance();
+  auto rrefForkData = ctx.prepareChildFork(rref);
```

Comment:
I know this is what PyRRef does, but as the comment there mentions this approach is error prone. If we jit serialize a model containing an RRef we would end up creating a child fork incorrectly here.

Can we ensure we create a child fork only when we're sending RPCs?

@mrshenli Do we have gh issue tracking that we need to fix this for PyRRef?

Reply:
Just created one #32966

Reply:
agree this is an issue we need to tackle. but I think it is better tackled with python pickling together, as in general JIT pickling serialization format should be same as Python pickling serialization format

Reply:
This is a pretty serious bug during pickling and unpickling. Can we fix this the right way in this PR?

Reply:
One option would be having a thread_local flag/context. We use this context/flag to wrap the code below (and for remote as well, basically every rpc related entry point). Then the pickler can check the thread_local context/flag. If not set, throw an error.

https://github.com/pytorch/pytorch/blob/d613bd0522d51602ccf8cb29d00eb26cf3af923d/torch/distributed/rpc/api.py#L457-L465

Reply:
Yanli and I spoke offline and thought its better to deal with this in a follow up PR. Although, lets fix this issue soon as a follow up PR for both jit serialization and PyRRef serialization.

The thread local idea sounds good since it can be reused across python pickle and JIT pickle.

Reply:
sounds good to me as well